### PR TITLE
[Engine] Add cooperative interruption

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
build on top of #16129

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/StoreBackedCommandExecutor.scala [hunk 1]
```diff
@@ -220,6 +221,14 @@ private[apiserver] final class StoreBackedCommandExecutor(
                 )
               )
             }
+
+        case ResultInterruption(continue) =>
+          resolveStep(
+            Tracked.value(
+              metrics.daml.execution.engineRunning,
+              trackSyncExecution(interpretationTimeNanos)(continue()),
```
Comment (remyhaemmerle-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/StoreBackedCommandExecutor.scala:229]: @mziolekda do I need to add a TODO ?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/LfValueTranslation.scala [hunk 2]
```diff
@@ -469,6 +469,9 @@ final class LfValueTranslation(
               )
               .map(resume)
               .flatMap(goAsync)
+
+          case LfEngine.ResultInterruption(continue) =>
+            goAsync(continue())
```
Comment (remyhaemmerle-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/LfValueTranslation.scala:474]: @mziolekda same think here

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 3]
```diff
@@ -393,7 +390,6 @@ class Engine(val config: EngineConfig = Engine.StableConfig) {
           handleError(err)
       }
 
-    @tailrec
```
Comment (akrmn) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:396]: I'm not familiar enough with scala but won't this have an impact on performance?

**Review by akrmn — APPROVED** [review]
Thanks @remyhaemmerle-da!

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 4]
```diff
@@ -359,7 +357,6 @@ class Engine(val config: EngineConfig = Engine.StableConfig) {
   }
 
   // TODO SC remove 'return', notwithstanding a love of unhandled exceptions
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:359]:
```suggestion
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:396]: revererted

Comment (akrmn) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:396]: again, I'm not familiar enough, but it seems that `loop` is now truly tail recursive in some cases, while in others it's recursive (but not tail recursive) by calling `interpretLoop`, so iiuc we'd lose stack safety

Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:396]: @tailrec grantees that  direct recursive calls do not use stack. Taking into account we have one such call (for `getTime`) dropping the annotation has a potential impact. If there are a lot of successive `getTime` call in the Daml code, the interpreter stack may blow up. Very unlikely but at least theoretically possible.

**Pull request merged**